buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy

```diff
@@ -103,16 +103,76 @@ class BuildPlugin implements Plugin<Project> {
+                // don't force transitive configurations to not conflict with themselves, since
+                // we just have them to find *what* transitive deps exist
+                return
+            }
+            configuration.resolutionStrategy.failOnVersionConflict()
+        })
 
         // force all dependencies added directly to compile/testCompile to be non-transitive, except for ES itself
-        project.configurations.compile.dependencies.all { dep ->
+        Closure disableTransitiveDeps = { ModuleDependency dep ->
             if (!(dep instanceof ProjectDependency) && dep.getGroup() != 'org.elasticsearch') {
                 dep.transitive = false
+
+                // also create a configuration just for this dependency version, so that later
+                // we can determine which dependencies it has (for pom generation excludes)
+                String depId = "${dep.getGroup()}:${dep.getName()}:${dep.getVersion()}"
+                String depConfig = "_transitive_${depId}"
+                if (project.configurations.findByName(depConfig) == null) {
+                    project.configurations.create(depConfig)
+                    project.dependencies.add(depConfig, depId)
+                }
             }
         }
-        project.configurations.testCompile.dependencies.all { dep ->
-            if (!(dep instanceof ProjectDependency) && dep.getGroup() != 'org.elasticsearch') {
-                dep.transitive = false
+
+        project.configurations.compile.dependencies.all(disableTransitiveDeps)
+        project.configurations.testCompile.dependencies.all(disableTransitiveDeps)
+        project.configurations.provided.dependencies.all(disableTransitiveDeps)
+
+        // add exclusions to the pom directly, for each of the transitive deps of this project's deps
+        project.modifyPom { MavenPom pom ->
+            pom.withXml { XmlProvider xml ->
+                // first find if we have dependencies at all, and grab the node
+                NodeList depsNodes = xml.asNode().get('dependencies')
+                if (depsNodes.isEmpty()) {
+                    return
+                }
+
+                // check each dependency for any transitive deps
+                for (Node depNode : depsNodes.get(0).children()) {
+                    String groupId = depNode.get('groupId').get(0).text()
+                    String artifactId = depNode.get('artifactId').get(0).text()
+                    String version = depNode.get('version').get(0).text()
+
+                    // collect the transitive deps now that we know what this dependency is
+                    String depId = "${groupId}:${artifactId}:${version}"
+                    String depConfig = "_transitive_${depId}"
+                    Configuration configuration = project.configurations.findByName(depConfig)
+                    if (configuration == null) {
+                        continue // we did not make this dep non-transitive
```

Comment:
We are only iterating direct dependencies here. The only way this configuration would not be present is if we decided not to create it above. So it must have been either a project or `org.elasticsearch`.